### PR TITLE
fix(db-performance): refactor messages mapper and replace queries with SQL View (AR-2435)

### PR DESCRIPTION
**Pull request by mchenani**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?
To achieve a better performance in fetching the messages of a conversation, we found some issues in the below parts of the code:
1- QUERYING DATA ON TOP OF QUERYs
We have different tables to store different message types, but when we were fetching them, we were querying those tables after we did a query for the messages list.
To fix this issue, we did some performance tests with different ideas; in the end, we found out that having a big join query to fetch the messages and all the related data would be a huge leap in the performance [ something between 10 to 13 times improvements 🤯 ha? Yes! 10 to 13 times!]
But still we needed that query in many other places and we needed it to be clean, the classic solution for this issue is the Views in SQL! We created a view in the messages.sq to represent this big query.

2- Mappers!
We were mapping the data after fetching the data via sqlDelight, but turns out sqlDelight support inner mappers and consumes them during the data fetch! which required some extra changes in the mapper.

3- More changes coming in other PR to contain more improvement.

Needs releases with:

- [x] GitHub link to other pull request -> update AR kalium ref.

#### Test Coverage (Optional)
No new tests added.
- [ ] I have added automated test to this contribution

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq [hunk 1]
```diff
@@ -124,17 +127,75 @@ CREATE TABLE MessageMissedCallContent (
...
+LEFT JOIN MessageTextContent AS TextContent ON Message.id = TextContent.message_id AND Message.conversation_id = TextContent.conversation_id
+LEFT JOIN MessageAssetContent AS AssetContent ON Message.id = AssetContent.message_id AND Message.conversation_id = AssetContent.conversation_id
+LEFT JOIN MessageMissedCallContent AS MissedCallContent ON Message.id = MissedCallContent.message_id AND Message.conversation_id = MissedCallContent.conversation_id
+LEFT JOIN MessageMemberChangeContent AS MemberChangeContent ON Message.id = MemberChangeContent.message_id AND Message.conversation_id = MemberChangeContent.conversation_id
+LEFT JOIN MessageUnknownContent AS UnknownContent ON Message.id = UnknownContent.message_id AND Message.conversation_id = UnknownContent.conversation_id
+LEFT JOIN MessageRestrictedAssetContent AS RestrictedAssetContent ON Message.id = RestrictedAssetContent.message_id AND RestrictedAssetContent.conversation_id = RestrictedAssetContent.conversation_id
+LEFT JOIN MessageFailedToDecryptContent AS FailedToDecryptContent ON Message.id = FailedToDecryptContent.message_id AND Message.conversation_id = FailedToDecryptContent.conversation_id;
```
Comment (vitorhugods) [inline persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq:186]: It's big, but performant tho.

Comment (mchenani) [reply persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq:186]: do you have any suggestions :D ?

Comment (vitorhugods) [reply persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq]: Not really. I'm just admiring it hahahahaha.

File: persistence/src/commonTest/kotlin/com/wire/kalium/persistence/dao/MessageDAOBenchmarkTest.kt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import com.wire.kalium.persistence.utils.stubs.newUserEntity
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.test.runTest
+import kotlin.random.Random
+import kotlin.test.BeforeTest
+import kotlin.test.Test
+import kotlin.time.ExperimentalTime
+import kotlin.time.measureTime
+
+class MessageDAOBenchmarkTest : BaseDatabaseTest() {
```
Comment (vitorhugods) [inline persistence/src/commonTest/kotlin/com/wire/kalium/persistence/dao/MessageDAOBenchmarkTest.kt:17]: Can we `@Ignore` this one?

If we start adding Benchmarks, CI will be super slow.

This one takes some seconds to run which slows down CI, and it's not something we will be monitoring anyway.

File: persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq [hunk 3]
```diff
@@ -21,6 +21,9 @@ CREATE TABLE Message (
       PRIMARY KEY (id, conversation_id)
 );
 
+-- Allows optimized querying of messages sorting by date.
+CREATE INDEX message_date_index ON Message(date);
```
Comment (yamilmedina) [inline persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq:25]: 👌

**Review by yamilmedina — COMMENTED** [review]
Looking super nice, we have some errors in tests, but should be simple to solve.
Also, would make sense (not now) to add a test  🤔 or maybe a comment on how should we maintain this view in case we add tables and/or columns, deleting should break anyway.

Awesome job! 🚀

**Review by vitorhugods — APPROVED** [review]
The sad reality is that the bigger the query, the faster it is :)

Right now we definitely need this. It's a bit annoying to add new fields, but we can figure it out and revisit in the future, taking advantage of the fact that we can easily turn the benchmark on and compare.

**Pull request merged**